Ansible: install kubernetes from github releases

zarezadeh opened the pull request:
This commit will enable the option of installing kubernetes components by downloading archives from github releases in ansible playbook. To enable this set `source_type: ghReleases` in `group_vars/all.yml`.
The default version of kubernetes is set to `v1.0.3` but it can be overridden by setting `kubernetes_version` variable

zarezadeh committed:
added ability to install kubernetes master and node components with downloading from github releases

eparis commented on ansible/roles/master/defaults/main.yml:
```diff
@@ -0,0 +1,5 @@
+---
+kubernetes_version: v1.0.3
```
Should be 1.0.4.  I wish github had a reasonable way to say 'latest'. Ugh

Arano-kai replied:
Found some solution for get latest release:
* Remove `kubernetes_version` from defaults
* Add task:
```
- name: Determine latest release
  shell: /bin/curl https://github.com/kubernetes/kubernetes/releases/latest 2>/dev/null | /bin/perl -n -e'/(https.*)(v\d([^\"]*))/ && print $2'
  when: kubernetes_version is undefined
```
* Add (commented out) `kubernetes_version` to group_vars for those, who wants some exact release.

***

About a task:
```
curl https://github.com/kubernetes/kubernetes/releases/latest
```
will get something like 
```
<html><body>You are being <a href="https://github.com/kubernetes/kubernetes/releases/tag/v1.0.4">redirected</a>.body></html>
```
with is redirect to latest release.
And
```
/bin/perl -n -e'/(https.*)(v\d([^\"]*))/ && print $2'
```
just regex to extract version from curl job.
I would prefer to use Python for this (since it is guarantee be in ansible managed nodes), but not familiar with it =/

eparis commented on ansible/roles/master/defaults/main.yml:
```diff
@@ -0,0 +1,5 @@
+kubernetes_version: v1.0.3
+kubernetes_download_filename: "kubernetes-{{ kubernetes_version }}.tar.gz"
+kubernetes_download_url: "https://github.com/kubernetes/kubernetes/releases/download/{{ kubernetes_version }}/kubernetes.tar.gz"
+kubernetes_download_dir: /usr/local/src
```
I dislike using `/usr`.  `/usr` is supposed to be for the distro. Do you have strong feelings against putting this in `/var` somewhere?

zarezadeh replied:
No, I don't have any preference toward `/usr` but I don't know which subdirectory in `/var` is more suitable. Maybe `/var/cache`?

eparis replied:
since they are temporary, maybe somewhere under /var/run/ ?

zarezadeh replied:
It seems /var/run is temporary too and it will be cleared after reboot

zarezadeh committed:
updated version of kubernetes to v1.0.4 in ansible playbook
added redundant variable in group_vars for `kubernetes_version` in order to be able to change node and master versions in one place

zarezadeh committed:
changed default location where kubernetes binaries will downloaded to

zarezadeh commented on ansible/roles/master/tasks/ghReleasesInstall.yml:
```diff
@@ -0,0 +1,42 @@
+- name: copy kubernetes master binaries
+  shell: cp {{ kubernetes_download_dir }}/kubernetes/server/bin/{{ item }} /usr/bin
+  args:
+    creates: "/usr/bin/{{ item }}"
```
With this we prevent unnecessary copying of binaries, but when updating to new version this won't copy new binaries. Anybody has a better approach?

eparis replied:
Maybe stat both src and dest with `get_checksum` and do the copy if they differ?

zarezadeh replied:
With multiple items, it's a little hard. Maybe we can leave it for now

zarezadeh replied:
Is it a good idea if I copy binaries with their name and version together like `kube-apiserver-v1.0.4` and then create symlink `kube-apiserver` that points to correct version?

eparis replied:
Lets unarchive into {{ kubernetes_download_dir}}/{{ kubernetes_version }}/  and instead of cp link from /usr/bin to the right binary.

zarezadeh replied:
I did this, but there was a minor issue. In master role there's a task that adds a linux capability to `kube-apiserver` and adding capabilities to links is not possible. I don't know enough about capabilities and don't know what's the best way for this scenario.

Arano-kai commented on ansible/group_vars/all.yml:
```diff
@@ -4,8 +4,12 @@
 source_type: localBuild
 
+# This version will be used only when installing kubernetes from github releases
+kubernetes_version: v1.0.4
```
Found some solution for get latest release:
* Remove `kubernetes_version` from defaults
* Add task:
```
- name: Determine latest release
  shell: /bin/curl https://github.com/kubernetes/kubernetes/releases/latest 2>/dev/null | /bin/perl -n -e'/(https.*)(v\d([^\"]*))/ && print $2'
  when: kubernetes_version is undefined
```
* Comment out `kubernetes_version` in group_vars for those, who wants some exact release.

***

About a task:
```
curl https://github.com/kubernetes/kubernetes/releases/latest
```
will get something like 
```
<html><body>You are being <a href="https://github.com/kubernetes/kubernetes/releases/tag/v1.0.4">redirected</a>.body></html>
```
with is redirect to latest release.
And
```
/bin/perl -n -e'/(https.*)(v\d([^\"]*))/ && print $2'
```
just regex to extract version from curl job.
I would prefer to use Python for this (since it is guarantee be in ansible managed nodes), but not familiar with it =/

zarezadeh replied:
I was able to grab latest version from github using standard shell tools:

```bash
curl -sI https://github.com/kubernetes/kubernetes/releases/latest | grep Location | rev | cut -d "/" -f 1 | rev
```
I'm going to try your solution with this and if successful add another commit.

zarezadeh committed:
made kubernetes_version optional, when unset will get latest stable release from github
moved kubernetes download dir to /var/cache/kube because /var/run is a tmpfs filesystem in most systems

eparis commented:
I'll merge if you rebase again.  So sorry....  Let me know when you rebase.

eparis commented:
I'm going to close. I'm sorry I ignored you for so long. If you re-open I promise to pay attention this time.
